pkg/util/push/request.go

```diff
@@ -0,0 +1,65 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package push
+
+import (
+	"fmt"
+
+	"github.com/grafana/mimir/pkg/mimirpb"
+)
+
+// supplierFunc should return either a non-nil body or a non-nil error. The cleanup function can be nil.
+type supplierFunc func() (*mimirpb.WriteRequest, func(), error)
+
+// Request represents a push request. It allows lazy body reading from the underlying http request
+// and adding cleanups that should be done after the request is completed.
+type Request struct {
+	cleanups []func()
+
+	getRequest supplierFunc
+
+	request *mimirpb.WriteRequest
+	err     error
+}
+
+func newRequest(p supplierFunc) *Request {
+	return &Request{
+		cleanups:   make([]func(), 0, 10),
```

Comment:
10 looks many cleanups. How many do we actually set in the worst case scenario?

Reply:
that was to future-proof it. 4 at the moment

Reply:
Btw, we can avoid one allocation here by using 1) array, 2) slice that array:
```
type struct Request {
  cleanupsArr [10]func()
  cleanups []func()
...
}

r.cleanups: cleanupsArr[:0]
```

That makes request itself bigger, but no extra allocation for backing array is required.

Reply:
neat. I pushed the change in https://github.com/grafana/mimir/pull/2419/commits/a0b383b7b1dc6bad34f992a15b85b5853e7c29ed